### PR TITLE
Item/Spell Cooldown Groups + Configurable Global Cooldown

### LINKED ISSUE #234: Consumables recharge time
**Is your feature request related to a problem? Please describe.**

I have a suggestion about the consumable's recharge time, in my game there are several types of foods. For example, in 2 seconds, I have about 5 types, one that fills 60 mana, another that fills 120 mana and another that fills 180 mana. When I use commode X, the recharge time only applies to food X, the other foods I can use, so the recharge time of 2 seconds is useless. Well, I can use all 5 foods in 1 second.

**Describe the solution you'd like**

You could select the consumables that will enter the recharge time at the same time.

**Additional context**

https://gyazo.com/31684d069130a19a5b027aac44f6b061

https://gyazo.com/304204fb8070200c92dc595c92f50af6

https://prnt.sc/snpb9r

### PR DESCRIPTION
**Pull request by Cheshire92**
resolves #234 

Adds a new field to items called a Cooldown Group. These groups can be configured through the item editor like you can folders.
image

Once set up, any item then sharing that group in a user's inventory will go on cooldown with their respective cooldowns upon usage of one of these items:
image3

Also added a little feature for a global cooldown, as well as a way to have specific items bypass it:
image4
Image5
Image6

Checklist:

- [x] Cooldown Groups for Items
- [x] Cooldown Groups for Spells
- [x] Configuration option for Spell and Item Cooldown Groups to overlap.
- [x] Configurable Global cooldown
- [x] Spell and Item option to bypass the Global Cooldown
- [X] Enable and disable functionality through the server configuration file.

### CODE REVIEW

**Review by jcsnider — APPROVED** [review]

**Comment by Azurebeats** [comment]
As discussed with @lodicolo , will be changing this to not be taken from the server config but a user enterable field stored as a string in the database instead for future localisation plans.

**Comment by jcsnider** [comment]
Code looks okay.. but there are a few bugs I think if I understand how this should work...

My setup:
Blue Mana Pot (10s CD)
Red Mana Pot (5s CD)

Scenario 1:
Use the blue mana pot with a 10s cd.
The red mana pot (visually) will only show a 5s cd timer.
notsynced

Scenario 2:
If you have a cd timer active when you pick up an item in the same cd group, the display shows the item as being ready for use, and it looks like the server actually adds the mana? (Visual bug client, logic bug server?)
[Video](https://www.ascensiongamedev.com/resources/filehost/ad716aa15e5e2e921f1d33f58859fd5c.mp4)

**Comment by Azurebeats** [comment]
Neither are bugs but how I wrote it.

I'll add a server option to copy the cooldown of the item used to each item in the group or the highest cooldown in the group rather than using the items respective cooldowns.

Could probably also write a fix to mark every item in a group to go on cooldown even if a player doesnt own the items.

**Comment by Azurebeats** [comment]
I've just changed the functionality to be more in line with what you described through configurable options @jcsnider 

https://github.com/AscensionGameDev/Intersect-Engine/pull/355/files#diff-6a0dfeb01db21d69a9d87c9aa128b5c0

File: Intersect.Editor/Forms/Editors/frmItem.cs [hunk 1]
```diff
@@ -926,7 +935,7 @@ public void InitEditor()
             cmbProjectile.Items.Add(Strings.General.none);
             cmbProjectile.Items.AddRange(ProjectileBase.Names);
 
-            //Collect folders
+            //Collect folders and colldown groups
```
Comment (lodicolo) [inline Intersect.Editor/Forms/Editors/frmItem.cs:938]: cooldown*

File: Intersect.Editor/Forms/Editors/frmItem.cs [hunk 2]
```diff
@@ -939,8 +948,19 @@ public void InitEditor()
...
+                if (!string.IsNullOrEmpty(((ItemBase)itm.Value).CooldownGroup) && 
+                    !mKnownCooldownGroups.Contains(((ItemBase)itm.Value).CooldownGroup))
+                {
+                    mKnownCooldownGroups.Add(((ItemBase)itm.Value).CooldownGroup);    
+                }
             }
 
+            mKnownCooldownGroups.Sort();
+            cmbCooldownGroup.Items.Clear();
+            cmbCooldownGroup.Items.Add("");
```
Comment (lodicolo) [inline Intersect.Editor/Forms/Editors/frmItem.cs:961]: `string.Empty`

File: Intersect.Server/Entities/Player.cs [hunk 3]
```diff
@@ -1957,19 +1957,56 @@ public void UseItem(int slot, Entity target = null)
                 if (itemBase.Cooldown > 0)
                 {
                     var cooldownReduction = 1 - this.GetCooldownReduction() / 100;
-                    if (ItemCooldowns.ContainsKey(itemBase.Id))
+                    // Are we dealing with a cooldown group, or an individual item?
+                    if (itemBase.CooldownGroup.Trim().Length > 0)
```
Comment (lodicolo) [inline Intersect.Server/Entities/Player.cs:1961]: `!string.IsNullOrWhiteSpace(itemBase.CooldownGroup)`

**Review by lodicolo — CHANGES_REQUESTED** [review]

File: Intersect.Editor/Forms/Editors/frmItem.cs [hunk 4]
```diff
@@ -939,8 +948,19 @@ public void InitEditor()
                         mKnownFolders.Add(((ItemBase) itm.Value).Folder);
                     }
                 }
+
+                if (!string.IsNullOrEmpty(((ItemBase)itm.Value).CooldownGroup) &&
```
Comment (lodicolo) [inline Intersect.Editor/Forms/Editors/frmItem.cs:952]: `IsNullOrWhiteSpace`

**Review by lodicolo — CHANGES_REQUESTED** [review]

**Comment by Cheshire92** [comment]
Added configurable cooldown groups to spells, as well as a config option to make them link/sync up between items and spells.
On top of that, I figured I'd add a rudimentary Global Cooldown system.

image

**Comment by Cheshire92** [comment]
And finished that as well..

Image1
Image2

**Review by lodicolo — DISMISSED** [review]

**Review by lodicolo — APPROVED** [review]

**Review by lodicolo — APPROVED** [review]

**Pull request merged**